oaknational/Oak-Web-Application
# Feat/bookmarks

## Description

- Adds browser-lib graphql files:
  - Bookmark queries/mutations
  - Codegen file which uses role "user" to get correct types
  - Script: `"gql-codegen:browser": "graphql-codegen --config src/browser-lib/graphql/codegen.yml -r dotenv/config",`
- Adds `useApolloClient` which listens for changes to `accessToken` and creates a new client when necessary 
- Adds Bookmarks:
  - context
  - provider
  - useBookmarks -> { bookmarks, addBookmark, removeBookmark, ... }
- Adds mocked apollo and bookmark providers to `renderWithProviders`
- Adds tests

## Issue(s)

Fixes #80 

## How to test

1. Go to https://samara-4ykhnzbcm-oak-national-academy.vercel.app/
3. Click "Sign in" in the top riht and go through the flow
4. Open [a lesson](https://samara-4ykhnzbcm-oak-national-academy.vercel.app/lessons/physics-only-review-chj3cd) or [another lesson](https://samara-4ykhnzbcm-oak-national-academy.vercel.app//lessons/lady-macbeth-c9jk2c) 
5. Click on "Add bookmark"
6. Go to the [bookmarks page](https://samara-4ykhnzbcm-oak-national-academy.vercel.app//bookmarks)
7. You should see your bookmarks
8. Click the "delete" button next to the bookmark
9. It should disappear
10. User state and bookmark state etc should update seemlessly across tabs 
11. Going on /bookmarks without being signed in should show you a message like below

## Screenshots

![image](https://user-images.githubusercontent.com/12934669/163013772-52521b73-19bf-41da-ad6d-7fce30b145fa.png)

![image](https://user-images.githubusercontent.com/12934669/163014204-39b0858f-7c5f-4493-85a7-c690cea42db2.png)
